src/pybind/mgr/affinity/module.py

```diff
@@ -0,0 +1,152 @@
+
+"""
+affinity module is aimed at managing multiple availability zone situations
+that need local read affinity.
+local read affinity in ceph means that all primary OSDs of a pool are from the same az.
+
+See doc/mgr/affinity.rst for more info.
+
+currently the only implemented feature is the ability to create such pool, by creating the correct CRUSH rule.
+in the futuer we will implement more features to monitor and manage failures, etc.
+"""
+
+from mgr_module import MgrModule, CommandResult, HandleCommandResult
+from threading import Event
+
+import os
+import json
+import subprocess
+from tempfile import mkstemp
+from mako.template import Template
+
+
+class Affinity(MgrModule):
+    COMMANDS = [
+        {
+            "cmd": "affinity create pool "
+                   "name=poolname,type=CephString,req=true "
+                   "name=region,type=CephString,req=true "
+                   "name=pd,type=CephString,req=true "
+                   "name=sd,type=CephString,req=true "
+                   "name=td,type=CephString,req=true",
+            "desc": "create a new pool with primary affinity to pd",
+            "perm": "rw"
+        },
+    ]
+
+    def __init__(self, *args, **kwargs):
+        super(Affinity, self).__init__(*args, **kwargs)
+
+        # set up some members to enable the serve() method and shutdown
+        self.run = True
+        self.event = Event()
+
+    def get_highest_rule_id(self):
+        crush = self.get('osd_map_crush')
+        crush_rules = crush['rules']
+        id = 1
+        for rule in crush_rules:
+            if rule['rule_id'] > id:
+                id = rule['rule_id']
+
+        return id
+
+    def create_crush_affined_rule(self, name, region, pd, sd, td):
+        rule_template = Template(""" rule ${name} {
```

Comment:
This is a pretty simple template... wouldn't ``format()`` be sufficient?  I think we should avoid adding new dependencies we don't need.

Reply:
yeha, probably,
at first I thought it'll be more complicated than that, I'll change.